sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs

```diff
@@ -29,9 +30,12 @@ private class Runner : IRunner
             /// continuously, asynchronously loop, waiting for these tasks to complete, and only
             /// exiting once the set becomes empty.
             /// </summary>
-            private readonly Dictionary<Task, List<string>> _inFlightTasks = new Dictionary<Task, List<string>>();
+            private readonly TaskMonitoringHelper _inFlightTasks = new TaskMonitoringHelper();
             private readonly List<Exception> _exceptions = new List<Exception>();
 
+            private readonly ConcurrentDictionary<(int TaskId, string Desc),int> _descriptions =
+                new ConcurrentDictionary<(int TaskId, string Desc),int>();
+
             public ImmutableList<Exception> SwallowedExceptions => this._exceptions.ToImmutableList();
```

Comment:
Curious if this list I put here and the `_exceptions` list are still necessary or if they can be lifted into your `TaskExceptionTracker` since that also contains a list of exceptions?

I guess this immutable list might still be necessary on the `IRunner` interface but maybe the `_exceptions` one can be lifted and the `TaskMonitoringHelper` can expose some property to access them?

Reply:
That's a good point, having a look. Looks like that list access is unsynchronized but coming from multiple threads so possibly another cause of woes. And.. yeah it looks like it could be the same list that I moved into the helper class. Let me unify.

Reply:
Ah actually no, TaskExceptionTracker it's adding and clearing it's internal buffer ~ the intent there was that it tries not to lose exceptions, but it temporarily holds them in between observing them in tasks and next time someone reads them via AwaitExceptionAsync. Exposing that list would make the interface to that class a bit more confusing.

The _exception list in Runner seems to be append-only, and I think Runner should continue to own that, if it's needed, I don't want to be tackling that one in this PR :)

I can however lock access to it so we don't race on it.

Reply:
Roger that, makes sense.